### DISCUSSION TITLE
How to get Grid to update when underlying data changes

### DISCUSSION THREAD

**Discussion by powertec-dan** [vikramlearning/blazorbootstrap#375] [Q&A] [answered]
I want to use a grid to show a data set.

This data set can change while the page is being displayed. I want to be able to reflect the changes in the Grid when the data changes.

I've already tried: InvokeAsync(() => StateChanged()); when the notification of the data change comes through and it does not update what is displayed in the Grid.

Any help appreciated.

Cheers,
Dan

**Comment by gvreddy04** [comment, marked as answer]
@powertec-dan Did you try the `await grid.RefreshDataAsync();`?

**Comment by powertec-dan** [comment]
Ah, my apologies, I did not see that in the documentation. I shall look more carefully in future.

Having implemented that, I'm now facing a crash when the collection is updated:

The code looks like this:

`    private void ItemsUpdated(List<Item>? items)
    {
        _items?.Clear();
        if (items != null)
            _items?.AddRange(items);
        grid?.RefreshDataAsync();
        InvokeAsync(() => StateHasChanged());
    }`

and the exception thrown is:

`More than one sibling of element 'tr' has the same value, '1'. Key values must be unique`

Rather than clearing the item list, I could go through and do deltas and update only what has changed. Would that help?

**Comment by gvreddy04** [comment]
@powertec-dan Could you please check if there are any duplicate rows?

**Reply by powertec-dan** [reply]
It suddenly occurs to me that this table has a composite primary key. 
Does the grid pick the first column as the PK? Is it possible to tell it to look at more than one column? 
Or do I need to create something like a hidden column that contains a guaranteed unique value?

**Reply by gvreddy04** [reply]
@powertec-dan It will take the **item** object, **HashCode**. I believe there are duplicate items. Can you add a unique property to the item object? Let me know if you need further help.

**Reply by powertec-dan** [reply]
Thank you!

That class _did_ implement HashCode (badly) and that was the reason. I've corrected it (and some other code around it like Equals) and it all works great now.

Thank you very much for your support. I very much appreciate it.

Cheers,
Dan
